doc/tools/docgram/doc_grammar.ml

```diff
@@ -0,0 +1,1349 @@
+open Coqpp_parser
+open Coqpp_ast
+
+let exit_code = ref 0
+
+let fprintf = Printf.fprintf
+
+let error s =
+  exit_code := 1;
+  Printf.eprintf "Error: ";
+  Printf.eprintf s
+
+let warn s =
+  Printf.eprintf "Warning: ";
+  Printf.eprintf s
+
+
+(* translated symbols *)
+
+type doc_symbol =
+| Sterm of string
+| Snterm of string
+| Slist1 of doc_symbol
+| Slist1sep of doc_symbol * doc_symbol
+| Slist0 of doc_symbol
+| Slist0sep of doc_symbol * doc_symbol
+| Sopt of doc_symbol
+| Sparen of doc_symbol list (* for GRAMMAR EXTEND *)
+| Sprod of doc_symbol list list (* for GRAMMAR EXTEND *)
+  (* editing operations *)
+| Sedit of string
+| Sedit2 of string * string
+
+(* nonterminals to rename or delete *)
+module StringMap = Map.Make(String)
+module NTMap = StringMap
+module StringSet = Set.Make(String)
+
+type gram = {
+  (* map from nonterminal name to a list of prods *)
+  (* each production is a list of doc_symbol *)
+  mutable map: doc_symbol list list NTMap.t;
+  (* specifies the order for nt's *)
+  mutable order: string list;
+}
+
+(*** Print routines ***)
+
+let sprintf = Printf.sprintf
+
+let map_and_concat f l =
+  String.concat "" (List.map f l)
+
+let rec output_prod plist need_semi = function
+    | Sterm s -> if plist then sprintf "%s " s else sprintf "\"%s\" " s
+    | Snterm s ->
+        if plist then sprintf "`%s` " s else
+          sprintf "%s%s " s (if s = "IDENT" && need_semi then ";" else "")
+    | Slist1 sym -> sprintf "LIST1 %s" (prod_to_str ~plist [sym])
+    | Slist1sep (sym, sep) -> sprintf "LIST1 %sSEP %s" (prod_to_str ~plist [sym]) (prod_to_str ~plist [sep])
+    | Slist0 sym -> sprintf "LIST0 %s" (prod_to_str ~plist [sym])
+    | Slist0sep (sym, sep) -> sprintf "LIST0 %sSEP %s" (prod_to_str ~plist [sym]) (prod_to_str ~plist [sep])
+    | Sopt sym -> sprintf "OPT %s" (prod_to_str ~plist [sym])
+    | Sparen sym_list -> sprintf "( %s) " (prod_to_str sym_list)
+    | Sprod sym_list -> sprintf "[ %s] "
+        (String.concat ""
+          (List.mapi (fun i r -> sprintf "%s%s" (if (i > 0) then "| " else "") (prod_to_str r)) sym_list))
+    | Sedit s -> sprintf "%s " s
+    (* todo: make PLUGIN output conditional on the set of prods? *)
+    | Sedit2 ("PLUGIN", s2) ->
+      if plist then
+        sprintf "     (%s plugin) " s2
+      else
+        sprintf "     (* %s plugin *) " s2
+    | Sedit2 (s, s2) -> sprintf "%s \"%s\" " s s2
+
+and prod_to_str_r plist prod =
+  match prod with
+  | p :: tl ->
+    let need_semi =
+      match prod with
+      | Snterm "IDENT" :: Sterm _ :: _
+      | Snterm "IDENT" :: Sprod _ :: _ -> true
+      | _ -> false in
+    (output_prod plist need_semi p) :: (prod_to_str_r plist tl)
+  | [] -> []
+
+and prod_to_str ?(plist=false) prod =
+  String.concat "" (prod_to_str_r plist prod)
+
+(* print a production *)
+let print_prod out plist nt prod first =
```

Comment:
This is suspect code. It seems you want to define a list-printing function, judging from the callers, in a quite intricate way. Please do that in one single function with pattern-matching on singleton cases to eliminate the `first` argument so as to make this more readable. You can take inspiration from the `Pp` list-acting functions.